apps/fft/fft_generator.cpp

```diff
@@ -0,0 +1,140 @@
+#include "Halide.h"
+
+#include "fft.h"
+
+namespace {
+
+using namespace Halide;
+
+enum class FFTNumberType { Real, Complex };
+std::map<std::string, FFTNumberType> fft_number_type_enum_map() {
+     return { { "real", FFTNumberType::Real },
+              { "complex", FFTNumberType::Complex } };
+}
+
+// Direction of FFT. Samples can be read as "time" or "spatial" depending
+// on the meaning of the input domain.
+enum class FFTDirection { SamplesToFrequency, FrequencyToSamples };
+std::map<std::string, FFTDirection> fft_direction_enum_map() {
+  return { { "samples_to_frequency", FFTDirection::SamplesToFrequency },
+           { "frequency_to_samples", FFTDirection::FrequencyToSamples } };
+}
+
+class FFTGenerator : public Halide::Generator<FFTGenerator> {
+public:
+
+    // Gain to apply to the FFT. This is folded into gains already
+    // being applied to the FFT. A gain of 1.0f indicates an
+    // unnormalized FFT. 1 / sqrt(N) gives a unitary transform 1/N
+    // gives a normalized transform for the forward direction. (Where
+    // "N" is the number of output elements of the FFT, the product of
+    // the dimensions for a complex FFT and the product of all the
+    // dimensions except the last and the last divided by 2 all minus
```

Comment:
I think it's confusing to mention the weird conjugate symmetry related stuff here. I think this is saying that the gain is a product of the dimensions of the output of the transform domain for a real FFT, which really isn't correct: To normalize a transform, the gain should be 1/N, where N is the number of samples in the input (no funky conjugate symmetry dimensions).